bindings/java/src/main/native/jucx_common_def.cc

```diff
@@ -16,14 +19,95 @@ static inline void log_error(const char* error)
+    jmethodID getAddress = env->GetMethodID(inetsockaddr_cls, "getAddress",
+                                            "()Ljava/net/InetAddress;");
+    jobject inet_address = env->CallObjectMethod(sock_addr, getAddress);
+
+    if (inet_address == NULL) {
+        JNU_ThrowException(env, "j2cInetSockAddr: InetSocketAddr.getAddress is null");
+        return false;
+    }
+
+    jclass inetaddr_cls = env->GetObjectClass(inet_address);
+
+    // Get address family. In Java IPv4 has addressFamily = 1, IPv6 = 2.
+    field = env->GetFieldID(inetaddr_cls, "holder",
+                            "Ljava/net/InetAddress$InetAddressHolder;");
+    jobject inet_addr_holder = env->GetObjectField(inet_address, field);
+    jclass inet_addr_holder_cls = env->GetObjectClass(inet_addr_holder);
+    field = env->GetFieldID(inet_addr_holder_cls, "family", "I");
+    jint family = env->GetIntField(inet_addr_holder, field);
+
+    field = env->GetStaticFieldID(inetaddr_cls, "IPv4", "I");
+    const int JAVA_IPV4_FAMILY = env->GetStaticIntField(inetaddr_cls, field);
+    field = env->GetStaticFieldID(inetaddr_cls, "IPv6", "I");
+    const int JAVA_IPV6_FAMILY = env->GetStaticIntField(inetaddr_cls, field);
+
+    // Get the byte array that stores the IP address bytes in the InetAddress.
+    jmethodID get_addr_bytes = env->GetMethodID(inetaddr_cls, "getAddress", "()[B");
+    jobject ip_byte_array = env->CallObjectMethod(inet_address, get_addr_bytes);
+
+    if (ip_byte_array == NULL) {
+        JNU_ThrowException(env, "j2cInetSockAddr: InetAddr.getAddress.getAddress is null");
+        return false;
+    }
+
+    jbyteArray addressBytes = static_cast<jbyteArray>(ip_byte_array);
+
+    if (family == JAVA_IPV4_FAMILY) {
+        // Deal with Inet4Address instances.
+        // We should represent this Inet4Address as an IPv4 sockaddr_in.
+        ss.ss_family = AF_INET;
+        sockaddr_in &sin = reinterpret_cast<sockaddr_in &>(ss);
+        sin.sin_port = htons(port);
+        jbyte *dst = reinterpret_cast<jbyte *>(&sin.sin_addr.s_addr);
+        env->GetByteArrayRegion(addressBytes, 0, 4, dst);
+        sa_len = sizeof(sockaddr_in);
+        return true;
+    } else if (family == JAVA_IPV6_FAMILY) {
+        jclass inet6_addr_cls = env->FindClass("java/net/Inet6Address");
+        ss.ss_family = AF_INET6;
+        sockaddr_in6& sin6 = reinterpret_cast<sockaddr_in6&>(ss);
+        sin6.sin6_port = htons(port);
+        // IPv6 address. Copy the bytes...
+        jbyte *dst = reinterpret_cast<jbyte *>(&sin6.sin6_addr.s6_addr);
+        env->GetByteArrayRegion(addressBytes, 0, 16, dst);
+        // ...and set the scope id...
+        jmethodID getScopeId = env->GetMethodID(inet6_addr_cls, "getScopeId", "()I");
+        sin6.sin6_scope_id = env->CallIntMethod(inet_address, getScopeId);
+        sa_len = sizeof(sockaddr_in6);
+        return true;
+    }
+    JNU_ThrowException(env, "Unknown InetAddress family");
+    return false;
```

Comment:
ok, so now this function always throws exceptions in case of error. Maybe we should make it return 'void', since there is no point of returning 'true'/'false' anymore?

Reply:
It throws an exception to java. But control flow will still be in c after.

Reply:
Throwing an exception in java from jni perspective is like just calling some java class method. There should be no uncatch exceptions in c, because it'll just kill JVM. Java exception can be intercepted in java.